File: app/src/main/java/com/agrogames/islandsofwar/engine/abs/common/Point.java
package com.agrogames.islandsofwar.engine.abs.common;

import androidx.annotation.NonNull;

import java.util.Objects;

public class Point {
    public float x;
    public float y;

    public Point(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public Point(Cell cell) {
        this.x = cell.x - 0.5f;
        this.y = cell.y - 0.5f;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Point point = (Point) o;
        return Float.compare(point.x, x) == 0 && Float.compare(point.y, y) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    public Point rotate(float angle){
        return new Point(
                (float) ((double)x * Math.cos(angle) - (double) y * Math.sin(angle)),
                (float) ((double)x * Math.sin(angle) + (double) y * Math.cos(angle))
        );
    }

    @NonNull
    @Override
    public String toString() {
        return "{" + x + ";" + y + "}";
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/engine/abs/common/Point.kt
package com.agrogames.islandsofwar.engine.abs.common

import java.util.Objects

class Point {
    var x: Float
    var y: Float

    constructor(x: Float, y: Float) {
        this.x = x
        this.y = y
    }

    constructor(cell: Cell?) {
        x = cell!!.x - 0.5f
        y = cell.y - 0.5f
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val point = o as Point
        return java.lang.Float.compare(point.x, x) == 0 && java.lang.Float.compare(point.y, y) == 0
    }

    override fun hashCode(): Int {
        return Objects.hash(x, y)
    }

    fun rotate(angle: Float): Point {
        return Point((x.toDouble() * Math.cos(angle.toDouble()) - y.toDouble() * Math.sin(angle.toDouble())).toFloat(), (x.toDouble() * Math.sin(angle.toDouble()) + y.toDouble() * Math.cos(angle.toDouble())).toFloat())
    }

    override fun toString(): String {
        return "{$x;$y}"
    }
}